### DISCUSSION TITLE
Problems with data I/O

### DISCUSSION THREAD

**Discussion by mauroarcidiacono** [neperfepx/neper#476] [Q&A] [answered]
Hello,

I am using Neper version 4.4.2-16 and gmsh version 4.7.1 to build a microstructure where only some grains have laths. To do it, I am using the following commands:

```
neper -T -n 30::from_morpho -id 4::1 -morpho "diameq:lognormal(1,0.35),1-sphericity:lognormal(0.145,0.03)::lamellar(w=file(lamwidth.txt),v=file(lamnormal.txt),pos=optimal)" -reg 1 -sel 0.08 -oridescriptor e -format tess,ori -o Polycrystal_laminated`

neper -M Polycrystal_laminated.tess -statelset vol -order 1 -cl 0.045 -meshqualexpr Odis -meshqualmin 1 -format inp,msh -o abq_input`
```

The microstructure generation works fine, but I am having trouble to do the following:

- How can I obtain the ID of the mesh elements that belong to each set? I need something like set1: el1 – el200, set2: el201 – el 500, etc.

- Is there a way to obtain the ID of all the laths? Is it possible to determine to which grain they belong? The laths are a different phase and I need their IDs to introduce their properties selectively. I tried using -group lam but it does not work (I get -1 for every structure). My lamwidth.txt file is constructed in the following way (I am only showcasing five lines):
        1 10.0
     	2 0.15:0.03
    	3 10.0
    	4 0.15:0.03
    	5 10.0
I am using this strategy to create laths in only some grains. Due to the shape of the grains and the laths, introducing a volumetric or size criteria would introduce errors in identifying the laths. I need a list with all the laths because they belong to a different phase. I also need ID ranges of the lamellar grains, for example, 2-8, where 3, 5 and 7 are thinner laths from a different phase.  

- Is it possible to introduce orientations to the laths of each grain? There is a crystallographic relationship between the grain and its laths and I would like to selectively input the orientation of the laths in those cases.

Thank you very much for your help.

**Comment by rquey** [comment, marked as answer]
Sorry for the delayed response. See inserts below.

> Hello,
> 
> I am using Neper version 4.4.2-16 and gmsh version 4.7.1 to build a microstructure where only some grains have laths. To do it, I am using the following commands:
> 
> ```
> neper -T -n 30::from_morpho -id 4::1 -morpho "diameq:lognormal(1,0.35),1-sphericity:lognormal(0.145,0.03)::lamellar(w=file(lamwidth.txt),v=file(lamnormal.txt),pos=optimal)" -reg 1 -sel 0.08 -oridescriptor e -format tess,ori -o Polycrystal_laminated`
> 
> neper -M Polycrystal_laminated.tess -statelset vol -order 1 -cl 0.045 -meshqualexpr Odis -meshqualmin 1 -format inp,msh -o abq_input`
> ```
> 
> The microstructure generation works fine, but I am having trouble to do the following:
> 
> How can I obtain the ID of the mesh elements that belong to each set? I need something like set1: el1 – el200, set2: el201 – el 500, etc.

The information in recorded in the mesh file, but another posibility is to use `-statelt elset3d`, which provides the elset each element belongs to. You can build the lists from there.

> Is there a way to obtain the ID of all the laths? Is it possible to determine to which grain they belong? The laths are a different phase and I need their IDs to introduce their properties selectively. I tried using -group lam but it does not work (I get -1 for every structure). My lamwidth.txt file is constructed in the following way (I am only showcasing five lines):
>       1 10.0
>       2 0.15:0.03
>       3 10.0
>       4 0.15:0.03
>       5 10.0
>       I am using this strategy to create laths in only some grains. Due to the shape of the grains and the laths, introducing a volumetric or size criteria would introduce errors in identifying the laths. I need a list with all the laths because they belong to a different phase. I also need ID ranges of the lamellar grains, for example, 2-8, where 3, 5 and 7 are thinner laths from a different phase.

`-group lam -statcell group` should work:
```
neper -T -n 2::from_morpho -morpho "voronoi::lamellar(w=0.1:0.2)" -group  lam -o 2lam
neper -V 2lam.tess -datacellcol group -print img
```
![image](https://user-images.githubusercontent.com/2659376/171208819-feea3a5a-49cc-4282-ac3e-06b56a238fd9.png)

To know which grain a lam belongs to, use `-statcell "scaleid(1)"`.
> 
> Is it possible to introduce orientations to the laths of each grain? There is a crystallographic relationship between the grain and its laths and I would like to selectively input the orientation of the laths in those cases.
> 
> Thank you very much for your help.

Yes, using `-ori`, a multiscale cell file and orientation files:
```
$ more my_msfile 
1 file(grain1.ori)
2 file(grain2.ori)
$ more grain1.ori 
   1.973863325448    8.075245608648    4.768505724249
  -0.308969168686   -0.587235038263    0.587731265354
  -0.559015418208    1.682899446703   -0.206136154737
   0.425754609764    1.303097365403   -0.482952202135
  -0.494697596547   -0.145357993821   -0.259166345035
  -0.031532633541    1.503020724248    2.958200923449
  -1.742213200329   -4.028203269231   -2.200636551085
   0.498777429009   -0.861308633139   -0.145046340636
   1.745674625290   -0.863340646663   -1.976604268526
$ more grain2.ori 
   0.919953402851    1.412193625247   -0.358468593221
  -1.381019825878    0.030115972514   -3.318384436434
   0.540501980937   -0.787561454202   -0.735523888182
  -2.005448934444    0.199749071961   -1.999283218069
   5.163148738562   14.893183177083    5.316486583072
   1.919036671923   -0.744035465279    0.303477757891
  -0.581246451853   -2.835847764762    0.780360761158
  -2.211174665124    0.445317073444    0.934247527182
  -2.194677973955   -0.312638828364   -0.439479476190
   1.153229166270   -1.695620636026    0.825414393959
```

You can set the orientations as follows, by re-running tessellation but this time using `-ori` so set orientations at scale 2:
```
$ neper -T -n 2::from_morpho -morpho "voronoi::lamellar(w=0.1:0.2)" -group lam -o 2lam -ori "random::msfile(my_msfile)"
```
Note that you need to know the number of lamellae per primary grain; so, you typically have to proceed in 2 steps.

**Comment by MaximePrs** [comment]
Hello,

> I tried using -group lam but it does not work (I get -1 for every structure)

I'm generating a dual phase material at a 2-scale tessalation. Then I need to identify each phase to add the corresponding properties in my modeling.
But I have a similar problem as above to identify the group of each grain. When using the `-group mode` command, I only get -1 for each structure.
I'm generating my structure like this:
`neper -T -n 10::40 -dim 2 -morpho "gg::diameq:0.7*lognormal(1.37,0.89)+0.3*lognormal(2.19,1.26)" -group mode -statcell id,group,area,circularity -o m_2D_perio_a_f`

I understand that in the case of lamellae, `-group lam` is working.
But is there a specific syntax to get the group identification of each grain with multiscale tessellation other than lamellae ?

**Reply by rquey** [reply]
Fixed in version 4.6.1-11.
